### PR TITLE
fix(providers): radancy stops trusting totalResults, cache-busts JSON, retries every fetch

### PR DESCRIPTION
**Pull request by ilyakanevskiy**
## Summary

Live-audited every Radancy (TalentBrew) tenant reachable from this session's `portals.yml` entry plus several found via web search, and found three independent, unrelated problems behind `radancy.mjs`'s `⚠️ truncated ... raise max_jobs/max_pages` warning: one tenant loses real postings to a caching bug, several tenants have a `totalResults` banner that simply overstates what their own search index serves on both transports, and one very large tenant hits a hard pagination-window ceiling at its backend. None of the three is fixed by raising `max_pages`/`max_jobs`, which is exactly the advice the current warning gives every time.

## Problem 1 — a caching bug drops and duplicates postings on Munich Re Group

Walking `careers.munichre.com`'s JSON fragment endpoint end-to-end and recording every `(page, id, title, location, url)` tuple showed the same posting reappearing on a second, distant page with byte-identical content, while a different posting never appeared on any page at all. Reproduced on 9 consecutive full walks with the same pages and the same magnitude every time (not random drift): 5 duplicated ids, always on the same page pairs.

Isolated the cause by testing two independent changes against the same tenant: `Cache-Control: no-cache` + `Pragma: no-cache` request headers made no difference (9/9 runs still broken), but a random per-request query parameter — forcing a cache-key miss — made it 0/9 broken. Re-running the same walk through the `?p=N` HTML transport (the same underlying data, a different route) never showed the problem at all, which further narrows it to a caching layer in front of the JSON route specifically, not the tenant's own data.

## Problem 2 — `totalResults` overstates reality on several tenants, and that's normal for this platform

Separately, four of nine fully-audited tenants (from a small board to Radancy's own corporate careers site) collected 10-56% fewer unique postings than their `data-total-results` banner claims, with zero duplicate rows anywhere. Re-walking two of them through the `?p=N` HTML transport reproduced the identical shortfall to the exact posting count, proving it isn't a transport bug — the tenant's own search index simply doesn't contain as many retrievable postings as its own counter claims. Cache-busting (the Problem 1 fix) made no difference to this shortfall on any of the affected tenants, confirming it's a separate issue.

`data-total-pages`, unlike `data-total-results`, matched the walk's own natural end (a short or empty final page) in every one of the 9 tenants checked, so it stays a valid bound for the walk itself — only the results counter is untrustworthy.

## Problem 3 — Walgreens hits a hard pagination-window ceiling at exactly 10,000

Walgreens (22,813 claimed postings, the largest tenant checked) was excluded from the count above and audited separately: a full 229-page walk shows pages 1-100 each returning a full 100 rows (10,000 total, no duplicates anywhere), then every page from 101 through 229 returning a clean, structured "no jobs" response (`hasJobs: false`, an empty results fragment) rather than an error or a short page. The cutoff lands exactly at offset 10,000 — 100 pages of 100 — the textbook default `max_result_window` ceiling on an Elasticsearch/Solr-style backend (one tenant observed so far, not established across the platform the way `workday.mjs`'s analogous `WORKDAY_OFFSET_CEILING` is). This isn't a caching artifact (a 130-page gap of identical empty responses isn't a stale cache hit) and it isn't recoverable by pagination cleverness on our side, unlike workday's clamp: TalentBrew's JSON fragment API exposes no documented facet-style split to route around it, so the backend simply refuses to serve past that offset no matter how the walk asks.

This surfaced a real gap while testing the fix live: with the *default* `max_jobs` (2,000), the walk correctly stops there first — page 20, well before the backend's own 10,000 ceiling — and now correctly warns (`truncated at 2,000 of 22,813 jobs`, `stopReason: 'cap'`), because raising `max_jobs` genuinely would recover more, up to 10,000. Only past that point does the *other* behavior apply: with `max_jobs` raised above 10,000, the walk runs to the backend's own empty page at 101 and returns exactly 10,000 with no warning, because past there raising anything no longer helps. Verified both live.

## The fix

- `buildFragmentUrl()` now appends a random `_` query parameter to every JSON fragment request, unique per call. Verified against all 11 live tenants seen in this investigation (including the one that already fails for an unrelated reason) that this changes nothing for a tenant that didn't have the caching problem — page-1 output matched with and without it in every case.
- The walk now tracks *why* it stopped (`stopReason`, matching the `workday.mjs`/`phenom.mjs` convention already used elsewhere in this file) instead of comparing the accumulated count against `totalResults`. The `raise max_jobs/max_pages` warning now fires only when `max_pages`/`max_jobs` demonstrably cut the walk short of the source's own larger page count — never when the walk ran to the source's own natural end and merely fell short of an inflated results counter, and never on a mid-walk fetch error (which now gets its own distinct, immediate log line instead of failing silently).
- `resolveMaxJobs()` now clamps a user-supplied `max_jobs` the same way `resolveMaxPages()` already clamps `max_pages` — noticed while testing Walgreens live: `max_pages` was already defended against a garbage config value, `max_jobs` wasn't. Not a resource-safety fix (the page loop was always independently bounded by `max_pages`, so an unclamped `max_jobs` never risked an unbounded walk) — just closing a defensive-coding gap between the two, capped at the true ceiling the JSON transport can ever return (`MAX_PAGES * FRAGMENT_RECORDS_PER_PAGE`).
- All three fetch call sites (the first JSON fragment request, every subsequent JSON page, the HTML `?p=N` fallback) now set `redirect:'error'` and are routed through `fetchJsonWithRetry`/`fetchTextWithRetry` instead of calling `ctx.fetchJson`/`ctx.fetchText` directly. Neither was true before this PR: `redirect:'error'` is `_http.mjs`'s per-provider-opt-in SSRF guard (#1440) that every peer provider sets and radancy.mjs didn't (flagged in review — DNS-layer `_ip-guard.mjs` already blocked a redirect to a private IP either way, so this closes a defense-in-depth gap, not a live bypass); the retry helper means a single transient 429/5xx/network blip on any page no longer permanently truncates the walk from that page on, the same protection every other multi-page provider already has.
- Fixed a false-positive `stopReason: 'cap'` flagged in review: `jobs.length >= maxJobs` alone doesn't prove `max_jobs` cut anything short — a tenant whose real total exactly fills the page(s) already walked (a naturally-complete result that happens to land exactly on `max_jobs`) was still classified as capped and warned about, even though nothing was left unfetched. Now `jobs.length > maxJobs` (a page overshot the buffer in one jump — real fetched rows the final slice still has to drop) caps unconditionally, and an exact match only counts when page budget was still available (`page <= lastPage`) when the loop stopped — i.e. `max_jobs` itself, not a coincidence of row count, is what kept a further page from being tried.
- `fetch()` now reads `ctx.maxPages` — set only by `verify-portals.mjs`'s bounded liveness probe, never during a real scan — to cap the walk to that budget (SHOULD, reference `providers/workday.mjs`) and to propagate any `ctx.fetch*` rejection unwrapped instead of absorbing it into the normal partial-result handling (MUST, reference `providers/vdab.mjs`). Before this PR `radancy.mjs` never read `ctx.maxPages` at all, and both per-page `catch` blocks would have swallowed `verify-portals.mjs`'s budget-exhaustion sentinel (`ProbePageBudgetReached`) into a normal `stopReason: 'error'`, misreporting a healthy tenant as broken during a probe instead of "live, partial". The cap-truncation warning is also suppressed while probing — the probe's own budget isn't the tenant's real config, and "raise max_pages" is not advice a liveness check has any use for.
- A non-string `json.results` on page 2+ (flagged in review) now throws a descriptive error instead of being silently coerced to `[]` and read as a natural end. Checked live before applying it: a genuine last page is still a string (Walgreens' own past-the-end response parses to zero rows but isn't itself empty or missing), so a non-string value has never actually been the documented "no more jobs" signal — coercing it to `[]` was silently reading a malformed response as the walk being done.

## Tenants checked

Munich Re Group is the tenant the original bug report was about; the rest were found via web search specifically to check whether the bug (and, separately, the fix) generalizes past that one case, not because career-ops is known to track them. None of this PR's tests or code changes depend on any of them being wired up anywhere.

All figures below are single, self-consistent snapshots from one full walk each — every tenant's own postings count changes over real time from ordinary hiring activity (confirmed separately: the same tenant's `totalResults` moved between checks made minutes apart), so figures taken at different points in this investigation are not expected to match each other and aren't compared across rows.

Table A — the caching bug (fixed by this PR), one before/after pair from two full walks:

| Company | URL | Before this fix | After this fix |
|---|---|---|---|
| Munich Re Group | https://careers.munichre.com/en/search-jobs | 1,572 collected of 1,579 claimed — 5 duplicated postings, always the same page pairs, reproduced on 9/9 consecutive full walks | 1,577 collected of 1,577 claimed — 0 duplicates, reproduced on 4/4 consecutive full walks |

Table B — `totalResults` overstates reality (unaffected by the caching fix; no longer produces a false "raise your limits" warning), one full walk each:

| Company | URL | totalResults | Actually collected (both transports agree) | Shortfall |
|---|---|---|---|---|
| Wegmans | https://jobs.wegmans.com/en | 554 | 442 | 20% |
| Baxter International | https://jobs.baxter.com/en | 610 | 549 | 10% |
| Club Med | https://www.clubmedjobs.com/en | 468 | 345 | 26% |
| Radancy (the vendor's own careers site) | https://jobs.radancy.com/en/search-jobs | 66 | 29 | 56% |

Table C — a hard backend pagination-window ceiling, distinct mechanism from Table B (confirmed via the exact page boundary, not inferred from the shortfall alone):

| Company | URL | totalResults | Actually collected | Where it stops |
|---|---|---|---|---|
| Walgreens | https://jobs.walgreens.com | 22,813 | 2,000 with the default `max_jobs`; 10,000 with `max_jobs` raised past the backend ceiling (e.g. 15,000) | Pages 1-100 full, page 101 onward a clean empty response (`hasJobs: false`) — the boundary sits exactly at offset 10,000, the standard search-backend `max_result_window` default |

Table D — confirmed clean, no bug of any kind (`totalResults` = postings actually collected, no duplicates), one full walk each:

| Company | URL | Postings |
|---|---|---|
| UnitedHealth Group (Optum) | https://careers.unitedhealthgroup.com/search-jobs | 5,593 |
| Kaiser Permanente | https://www.kaiserpermanentejobs.org/search-jobs | 2,961 |
| Cargill | https://careers.cargill.com | 1,407 |
| Comcast | https://jobs.comcast.com | 689 |

Table E — an unrelated finding from the same search, kept only because it's an instructive edge case; not a radancy.mjs bug and not fixed by this PR:

| Company | URL | Issue |
|---|---|---|
| AT&T | https://www.att.jobs/ | `radancy.mjs` makes no preliminary request — the JSON fragment fetch (`buildFragmentUrl`, the exact URL it calls first) is itself the one that fails. Its `Content-Security-Policy` and `Feature-Policy` response headers are each ~10.6 KB, past the ~16 KB total-header limit Node enforces by default — `fetch` (and the raw `https` module, confirmed with the same error) both refuse the response with `HeadersOverflowError`/`Parse Error: Header overflow` before any provider code runs, on that exact fragment URL, not just the surrounding HTML page (checked both directly). This is process-wide, not fetch- or undici-specific, and it's an environment setting, not a code bug: `NODE_OPTIONS=--max-http-header-size=65536` (or the equivalent `--max-http-header-size` CLI flag) resolves it — verified live. Confirmed the failure as-is is caught cleanly by `radancy.mjs`'s existing `try/catch` and does not crash the process; the provider correctly falls through and ultimately reports the board unreachable rather than silently empty. Not touched in this PR since it isn't a radancy-specific fix — a global launch-time setting belongs with however career-ops's entrypoints get invoked, not inside one provider. |

## Verification

- `node test-all.mjs --only providers/` — 2,114 passed, 0 failed. Added 10 new assertions: the cache-buster is present and unique per call, no warning fires when a natural page-1 end falls short of an inflated `totalResults`, a genuine `max_pages` cap (source reports more pages than our ceiling) still warns, an exact `max_jobs` match on a naturally-complete result does not false-positive into a cap warning, a mid-walk JSON fetch error keeps earlier pages and logs its own message without cap advice, `redirect:'error'` is set on every JSON and every HTML request, a page that fails once transiently recovers via retry with no data lost and no warning, and `resolveMaxJobs()` clamps an absurd value and still defaults correctly.
- Re-ran the fixed provider live against Munich Re Group 4 times: 1,577 jobs, 1,577 unique URLs, no warning, every time.
- Re-ran it live against Wegmans: 442 jobs, no warning (previously would have logged a "raise max_jobs/max_pages" warning that could not have fixed anything).
- Re-ran it live against Walgreens (Table C) twice: with the default `max_jobs`, 2,000 jobs and a correct, actionable `⚠️` cap warning (`stopReason: 'cap'` — `max_jobs` genuinely is what's limiting it, up to the backend's own 10,000); with `max_jobs: 15000`, exactly 10,000 jobs and no warning — the walk reaches the backend's own empty page at 101 and correctly reads that as a natural stop, not a local cap.

## Not in this PR

- AT&T's header-overflow fetch failure (Table E) — resolvable with a Node process launch setting (`--max-http-header-size`/`NODE_OPTIONS`), not a pagination/caching fix, so it doesn't belong inside `radancy.mjs`.
- Walgreens' 10,000-result backend ceiling (Table C) isn't something career-ops can work around from the client side — the source itself won't serve more no matter how it's asked. Nothing to fix beyond correctly not blaming a local limit for it, which this PR already does.
- The `?p=N` HTML fallback loop has no equivalent `max_pages` cap warning at all today (it silently stops at the ceiling). Out of scope here since it wasn't part of what this investigation verified live; worth a follow-up.
- A host allowlist for `resolveListUrl()` was suggested in review (CodeRabbit) but declined: unlike `greenhouse.mjs`/`workday.mjs`, radancy has no fixed vendor domain to match against — it's designed for arbitrary customer-branded hosts, wired explicitly per entry (`provider: radancy`, no `detect()`). `phenom.mjs`, the same explicit-wiring/branded-host architecture, has the identical shape (`u.protocol !== 'https:' && u.protocol !== 'http:'`, no allowlist) — this repo's actual SSRF protection for that architecture is the central DNS/IP-layer guard (`_ip-guard.mjs`, #3096), already applied to every fetch regardless. Matches the code-owner review's own earlier assessment on this PR.

### CODE REVIEW

**Review by Scott-Emberson — COMMENTED** [review]
Verified the change this PR actually makes (stop trusting `totalResults`, cache-bust the fragment requests), and it is sound and well-pinned. The walk is bounded by `Math.min(totalPages ?? maxPages, maxPages)` plus natural stops (empty page, all-duplicate page); `totalResults` is display-only and the cap warning is gated on `stopReason==='cap'`, never on `returned < totalResults`. The owned test reddens on regression: an inflated `totalResults` with a natural page-1 end must produce zero warnings (restoring the old `returned < totalResults` warn reddens it), a genuine `max_pages < totalPages` cap warns, a mid-walk error keeps the earlier page, and two `buildFragmentUrl` calls must yield different cache-buster values (reddens if `randomUUID()` is made deterministic).

One real finding on the SSRF axis, and I want to flag it rather than decide it unilaterally. This repo guards SSRF centrally at the DNS/IP layer (`_ip-guard.mjs`, #3096, which wraps the whole fetch including redirect targets), so there is no per-provider host allowlist by design. But `radancy.mjs`'s `fetchJson`/`fetchText` call sites omit `redirect:'error'`, which `_http.mjs`'s own comment calls "every provider's mandatory SSRF guard (#1440)" and which peer providers (lever, jibeapply, csod) all set. A redirect to a private IP is still blocked at DNS resolution, so this is defense-in-depth rather than a live bypass, and it is pre-existing (the JSON transport predates this PR, which did not touch it). There is also no spoof-host test asserting zero fetches, though given the central guard there is no per-provider allowlist for such a test to exercise.

Because this is a provider in the code-owner area but the deviation is pre-existing and mitigated, I am not blocking this PR's own change over it, but I am also not signing it off clean. My recommendation is to add `redirect:'error'` to both radancy fetch call sites to match the #1440 convention. Deferring the block-or-accept call to the maintainer. The `totalResults`/cache-bust change itself is verified sound.

**Review by Scott-Emberson — COMMENTED** [review]
Thanks, this closes the point I raised. `redirect:'error'` is now on all three transport call sites (both fragment JSON fetches and the HTML `?p=N` fallback), which is every fetch radancy makes (there is no per-job detail fetch), so the #1440 deviation is gone. The owned test now pins it too: it asserts `redirect` is `'error'` on every fragment call and every HTML call, so a future removal on any site reddens. Good layer for it.

I re-verified the rest at this head:

- The stop-trusting-`totalResults` change stays sound and is not weakened by the retry addition. The walk is bounded by `Math.min(totalPages ?? maxPages, maxPages)` plus natural empty/duplicate-page stops, `totalResults` is display-only and never a stop or warn input (the exact-fit and inflated-total cases assert no warning, the genuine cap case does warn), and the cache-buster uniqueness is asserted.
- The retry addition is safe: all radancy requests are GET, so no non-idempotent retry; it is bounded to three attempts; and it explicitly does not retry a refused-redirect TypeError, so the SSRF guard short-circuits rather than burning attempts. Transient-recover, permanent-503-partial-preservation, and 404-non-retryable are all tested.

Owned test verified sound, the SSRF-hygiene deviation is resolved, and CI is green with the branch not behind. This is GATE-tier and now clean. Holding formal approval until Scott gives the go-ahead, and surfacing it to him as ready-to-approve.

File: providers/radancy.mjs [hunk 1]
```diff
@@ -303,34 +353,68 @@ export default {
...
+          }
+          // The loop only reaches here without an early break when it walked
+          // every page up to `lastPage`. That is only OUR cap, not the
+          // source's own end, when `lastPage` is `maxPages` and the source
+          // either gave no page count or claimed more pages than our ceiling.
+          if (stopReason === 'complete' && page > lastPage && lastPage === maxPages
+            && (totalPages == null || totalPages > maxPages)) {
+            stopReason = 'cap';
           }
+          if (jobs.length >= maxJobs) stopReason = 'cap';
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/radancy.mjs]: Fixed, close to your suggested diff: `jobs.length > maxJobs` now unconditionally caps (a page pushing the buffer strictly past maxJobs in one jump still means real, already-fetched rows get sliced off at the return, even when that page turns out to be the true final one — page <= lastPage would have wrongly excused that case), and the exact-match case (`jobs.length === maxJobs`) stays gated on `page <= lastPage` as you proposed. Added the requested exact-max_jobs-natural-end test.

File: providers/radancy.mjs [hunk 2]
```diff
@@ -303,34 +354,68 @@ export default {
...
             await wait(PAGE_DELAY_MS);
             let rows;
             try {
-              const json = await ctx.fetchJson(buildFragmentUrl(listUrl, page), {
+              const json = await fetchJsonWithRetry(ctx, buildFragmentUrl(listUrl, page), {
+                redirect: 'error',
                 headers: { accept: 'application/json', 'x-requested-with': 'XMLHttpRequest' },
               });
               const frag = typeof json?.results === 'string' ? json.results : '';
               rows = frag ? parseResults(frag, origin) : [];
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/radancy.mjs]: Fixed as suggested. Checked live before applying it: a genuine last page is still a STRING (Walgreens' own past-the-end response answers `hasJobs:false` with a non-empty shell string that simply parses to zero rows, not an empty or missing `results`), so a non-string `results` has never been the observed "no more jobs" signal — coercing it to `[]` was silently reading a malformed response as a natural end. Now throws descriptively and keeps earlier pages, with a test.

**Review by Scott-Emberson — COMMENTED** [review]
Re-verified at the force-pushed head. The rewrite kept all four gate properties: redirect:'error' is still on all three fetch sites and still asserted per-transport by the owned test, totalResults stays display-only (walk bounded by min(totalPages ?? maxPages, maxPages), warn gated on stopReason==='cap', cache-buster uniqueness asserted), the retry is still bounded to three attempts, GET-only, and skips a refused-redirect TypeError, and the malformed-response and probe-sentinel false-pass guards are intact (a non-string page-2 results throws rather than faking a natural end). The owned suite passes at this head and the branch is not behind. Still sound and clean; ready-to-approve, holding for Scott's go-ahead.
